### PR TITLE
added avatar generator from user initials

### PR DESCRIPTION
**Pull request by rezk2ll**
example of generated avatar for the `John` user:

image

### CODE REVIEW

**Review by oubchid — DISMISSED** [review]

File: src/frontend/js/modules/avatar.js [hunk 1]
```diff
@@ -574,6 +574,91 @@
...
+        );
+        context.fill();
+
+        context.font = (canvas.height / 3) + 'px serif';
+        context.fillStyle = '#000';
+        context.textAlign = 'center';
+        context.textBaseline = 'middle';
+        context.fillText(text, canvas.width / 2, canvas.height / 2);
+
+        return canvas.toDataURL();
```
Comment (oubchid) [inline src/frontend/js/modules/avatar.js:680]: Maybe we can reduce image quality, this could perhaps improve the loading  ?
`canvas.toDataURL('image/jpeg', 0.5);`

**Review by oubchid — DISMISSED** [review]

File: src/frontend/js/modules/avatar.js [hunk 2]
```diff
@@ -574,6 +574,91 @@
...
+       *
+       * @param {String} name - the username or email
+       * @returns {String} - the color
+       */
+      function generateColor(seed) {
+        let hash = 0;
+
+        seed.split('').forEach(c => {
+          // eslint-disable-next-line no-bitwise
+          hash = ((hash << 5) - hash) + c.charCodeAt(0);
```
Comment (alagane) [inline src/frontend/js/modules/avatar.js:625]: I don't get this part 🤔

File: src/frontend/js/modules/avatar.js [hunk 3]
```diff
@@ -574,6 +574,91 @@
...
+        });
+
+        // eslint-disable-next-line no-bitwise
+        const color = (hash & 0x00FFFFFF)
+          .toString(16)
+          .toUpperCase();
```
Comment (alagane) [inline src/frontend/js/modules/avatar.js]: It is not needed to uppercase it to make a valid color string ;)

File: src/frontend/js/modules/avatar.js (hunk 2 above)
Comment (rezk2ll) [reply src/frontend/js/modules/avatar.js:625]: creates a unique hash for a string, the hash will be the same for that string, I chose this method to avoid random colors for the same name.

File: src/frontend/js/modules/avatar.js (hunk 1 above)
Comment (rezk2ll) [reply src/frontend/js/modules/avatar.js:680]: we don't load anything from the network or whatsoever, it's just a base64 string generated from a service that exists within the frontend app itself.

File: src/frontend/js/modules/avatar.js (hunk 2 above)
Comment (alagane) [reply src/frontend/js/modules/avatar.js:625]: Why shift it by 5, and not 8, or 16? Why substract it after?

Comment (rezk2ll) [reply src/frontend/js/modules/avatar.js:625]: detailed explanation:
https://www.javamex.com/tutorials/collections/hash_function_technical_2.shtml

tldr;
`hash << 5` is the same as `hash * 32` and `((hash << 5) - hash)` is the same as `(hash * 31)`
the reason behind using multiplication by `31` is because it's faster.

> The value 31 was chosen because it is an odd prime. If it were even and the multiplication overflowed, information would be lost, as multiplication by 2 is equivalent to shifting. The advantage of using a prime is less clear, but it is traditional. A nice property of 31 is that the multiplication can be replaced by a shift and a subtraction for better performance: 31 * i == (i << 5) - i. Modern VMs do this sort of optimization automatically.

(from Chapter 3, Item 9: Always override hashcode when you override equals, page 48)

**Review by alagane — DISMISSED** [review]
